### ISSUE TITLE
`discarded_futures` : Unwaited return for when Future is expected

### ISSUE THREAD

**Issue by bsz0206** [dart-lang/sdk#59331]
Unwaited does work only on void returns. We are forced to ignore, which is not nice:

FutureBuilder(
    // ignore: discarded_futures
    future: asyncCall(),

Unwaited does work only on void returns:
FutureBuilder(
    future: unawaited(asyncCall()),,

**Comment by FMorschel** [comment]
Hey @bsz0206, considering https://github.com/dart-lang/sdk/issues/59204, I've tested 

```dart
import 'dart:async';

Future<void> f() async {}

void g() {
  unawaited(f());
}
```

And got no lint, so I think this can be closed?

---

If you have any other problems with the `FutureBuilder` widget, you may want to have a look at (which are all duplicates):

- https://github.com/dart-lang/sdk/issues/59504
- https://github.com/dart-lang/sdk/issues/58889
- https://github.com/dart-lang/sdk/issues/59455
- https://github.com/dart-lang/sdk/issues/59151
- https://github.com/dart-lang/sdk/issues/59387

**Comment by FMorschel** [comment]
Hey @bsz0206, considering https://github.com/dart-lang/sdk/issues/59204, I've tested 

```dart
import 'dart:async';

Future<void> f() async {}

void g() {
  unawaited(f());
}
```

And got no lint, so I think this can be closed?

---

If you have any other problems with the `FutureBuilder` widget, you may want to have a look at (which are all duplicates):

- https://github.com/dart-lang/sdk/issues/59504
- https://github.com/dart-lang/sdk/issues/58889
- https://github.com/dart-lang/sdk/issues/59455
- https://github.com/dart-lang/sdk/issues/59151
- https://github.com/dart-lang/sdk/issues/59387

**Comment by bsz0206** [comment]
No, the use case is different. The future functions returns something other than void:

Future<String> getMeAnString() async {
/* ... */ 
}

FutureBuilder(
       future: getMeAnString(),
       builder: ..
       ...
)

then I would get: Asynchronous function invoked in a non-'async' function. Try converting the enclosing function to be 'async' and then 'await' the future. and I will have to add an //ignore :

FutureBuilder(
       // ignore: discarded_futures
       future: getMeAnString(),
       builder: ..
       ...
)

This literally means almost every FutureBuilder would need an extra ignore that is not nice.

> Hey @bsz0206, considering #59204, I've tested
> ```dart
> import 'dart:async';
> 
> Future<void> f() async {}
> 
> void g() {
>   unawaited(f());
> }
> ```
> 
> And got no lint, so I think this can be closed?

**Comment by FMorschel** [comment]
@pq here is the CL: https://dart-review.googlesource.com/c/sdk/+/403901

Once you review it, can you also take a look at all of the issues mentioned to make sure you agree they are dupes? Thanks!

**Comment by FMorschel** [comment]
I noticed that this should consider `FutureOr`. I'll review it later today. Also, I'll open another issue to address that it shows the `unnawaited` fix for `FutureOr` but that is not valid.
